musileko/client/gl/gl.zig

```diff
@@ -23,4 +29,64 @@ pub fn enableDepthTest() void {
 
 pub fn disableDepthTest() void {
     c.glDisable(c.GL_DEPTH_TEST);
-}
\ No newline at end of file
+}
+
+const std = @import("std");
+const panic = std.debug.panic;
+
+const gl_debug_log = std.log.scoped(.gl_debug);
+
+pub fn gl_debug_callback(
+    source: c_uint,
+    debugtype: c_uint,
+    id: c_uint,
+    severity: c_uint,
+    length: c_int,
+    message: [*c]const u8,
+    user_param: ?*const anyopaque,
+) callconv(.C) void {
+    _ = length;
+    _ = user_param;
+
+    const source_str = switch (source) {
+        c.GL_DEBUG_SOURCE_API => "API",
+        c.GL_DEBUG_SOURCE_APPLICATION => "Application",
+        c.GL_DEBUG_SOURCE_OTHER => "Other",
+        c.GL_DEBUG_SOURCE_SHADER_COMPILER => "Compiler",
+        c.GL_DEBUG_SOURCE_THIRD_PARTY => "Third Party",
+        c.GL_DEBUG_SOURCE_WINDOW_SYSTEM => "Window System",
+        else => panic("Invalid GL debug source {}", .{ source }),
+    };
+
+    const debugtype_str = switch (debugtype) {
+        c.GL_DEBUG_TYPE_DEPRECATED_BEHAVIOR => "Deprecated Behavior",
+        c.GL_DEBUG_TYPE_ERROR => "Error",
+        c.GL_DEBUG_TYPE_MARKER => "Marker",
+        c.GL_DEBUG_TYPE_OTHER => "Other",
+        c.GL_DEBUG_TYPE_PERFORMANCE => "Performance",
+        c.GL_DEBUG_TYPE_POP_GROUP => "Pop Group",
+        c.GL_DEBUG_TYPE_PORTABILITY => "Portability",
+        c.GL_DEBUG_TYPE_PUSH_GROUP => "Push Group",
+        c.GL_DEBUG_TYPE_UNDEFINED_BEHAVIOR => "Undefined Behavior",
+        else => panic("Invalid GL debug type {}", .{ debugtype }),
+    };
+
+    switch (severity) {
+        c.GL_DEBUG_SEVERITY_NOTIFICATION => {
+            if (source != c.GL_DEBUG_SOURCE_SHADER_COMPILER) {
+                // ignore shader compiler notifications by default as it produces a good number of them and they're not super useful
+                gl_debug_log.info("ID: {X}, Source: {s}, Type: {s}, Message: {s}", .{id, source_str, debugtype_str, message});
+            }
+        },
+        c.GL_DEBUG_SEVERITY_LOW => {
+            gl_debug_log.debug("ID: {X}, Source: {s}, Type: {s}, Message: {s}", .{id, source_str, debugtype_str, message});
+        },
+        c.GL_DEBUG_SEVERITY_MEDIUM => {
+            gl_debug_log.warn("ID: {X}, Source: {s}, Type: {s}, Message: {s}", .{id, source_str, debugtype_str, message});
+        },
+        c.GL_DEBUG_SEVERITY_HIGH => {
+            panic("ID: {X}, Source: {s}, Type: {s}, Message: {s}", .{id, source_str, debugtype_str, message});
```

Comment:
This doesn't need to be a panic, but errors of this type are things like GL_INVALID_OPERATION, where something has definitely gone wrong, so while it's *technically* recoverable, it's also a high likelihood that the code is doing something flat out wrong

I also forgot to change it back to logging an error before committing